### ISSUE TITLE
Feature: Setup CI/CD for nuget

### ISSUE THREAD

**Issue by ProphetLamb** [ProphetLamb/Surreal.Net#7]
Building and testing should be automated using GitHub actions, appveyor, or whatever

**Comment by AntoniosBarotsis** [comment]
I could put this together! Do you just want a simple build and test or anything more specific?

**Comment by AntoniosBarotsis** [comment]
I'm having some issues getting the tests to run.

I tried running the DB through compose (which seems to exit immediately) and `docker run --rm -p 8082:8000 surrealdb/surrealdb:latest start` as listed [here](https://surrealdb.com/install) but some tests are failing, is this supposed to happen?


```
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:00.72]     Surreal.Net.Tests.DatabaseTests.RestTestSuite2 [FAIL]
[xUnit.net 00:00:00.78]     Surreal.Net.Tests.DatabaseTests.RpcTestSuite [FAIL]
[xUnit.net 00:00:00.87]     Surreal.Net.Tests.RpcClientTests.Signin [FAIL]
[xUnit.net 00:00:00.88]     Surreal.Net.Tests.RpcClientTests.Create [FAIL]
[xUnit.net 00:00:00.90]     Surreal.Net.Tests.RpcClientTests.Change [FAIL]
  Failed Surreal.Net.Tests.DatabaseTests.RestTestSuite2 [40 ms]
  Error Message:
   System.Exception : Expected Ok, got 1 (: ) in useResp
  Stack Trace:
     at Surreal.Net.Tests.DatabaseTestDriver`2.AssertOk(ISurrealResponse& rpcResponse, String caller) in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\DatabaseTests.cs:line 135
   at Surreal.Net.Tests.DatabaseTestDriver`2.Run() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\DatabaseTests.cs:line 43
   at Surreal.Net.Tests.DatabaseTests.RestTestSuite2() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\DatabaseTests.cs:line 19
--- End of stack trace from previous location ---
  Failed Surreal.Net.Tests.DatabaseTests.RpcTestSuite [65 ms]
  Error Message:
   System.Exception : Expected Ok, got -32000 (There was a problem with authentication) in signInStatus
  Stack Trace:
     at Surreal.Net.Tests.DatabaseTestDriver`2.AssertOk(ISurrealResponse& rpcResponse, String caller) in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\DatabaseTests.cs:line 135
   at Surreal.Net.Tests.DatabaseTestDriver`2.Run() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\DatabaseTests.cs:line 53
   at Surreal.Net.Tests.DatabaseTests.RpcTestSuite() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\DatabaseTests.cs:line 13
--- End of stack trace from previous location ---
  Failed Surreal.Net.Tests.RpcClientTests.Signin [227 ms]
  Error Message:
   Expected rsp.Error to be <null>, but found Surreal.Net.RpcError
{
    Code = -32000,
    Message = "There was a problem with authentication"
}.
  Stack Trace:
     at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Primitives.ReferenceTypeAssertions`2.BeNull(String because, Object[] becauseArgs)
   at Surreal.Net.Tests.RpcClientTests.Signin() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 27
--- End of stack trace from previous location ---
  Failed Surreal.Net.Tests.RpcClientTests.Create [14 ms]
  Error Message:
   Expected rsp.Error to be <null>, but found Surreal.Net.RpcError
{
    Code = -32000,
    Message = "There was a problem with authentication"
}.
  Stack Trace:
     at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Primitives.ReferenceTypeAssertions`2.BeNull(String because, Object[] becauseArgs)
   at Surreal.Net.Tests.RpcClientTests.Signin() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 27
   at Surreal.Net.Tests.RpcClientTests.Use() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 34
   at Surreal.Net.Tests.RpcClientTests.Create() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 51
--- End of stack trace from previous location ---
  Failed Surreal.Net.Tests.RpcClientTests.Change [14 ms]
  Error Message:
   Expected rsp.Error to be <null>, but found Surreal.Net.RpcError
{
    Code = -32000,
    Message = "There was a problem with authentication"
}.
  Stack Trace:
     at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Primitives.ReferenceTypeAssertions`2.BeNull(String because, Object[] becauseArgs)
   at Surreal.Net.Tests.RpcClientTests.Signin() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 27
   at Surreal.Net.Tests.RpcClientTests.Use() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 34
   at Surreal.Net.Tests.RpcClientTests.Create() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 51
   at Surreal.Net.Tests.RpcClientTests.Change() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 82
--- End of stack trace from previous location ---
[xUnit.net 00:00:00.95]     Surreal.Net.Tests.RpcClientTests.Query [FAIL]
[xUnit.net 00:00:00.95]     Surreal.Net.Tests.RpcClientTests.Select [FAIL]
[xUnit.net 00:00:00.95]     Surreal.Net.Tests.RpcClientTests.Use [FAIL]
  Failed Surreal.Net.Tests.RpcClientTests.Query [9 ms]
  Error Message:
   Expected rsp.Error to be <null>, but found Surreal.Net.RpcError
{
    Code = -32000,
    Message = "There was a problem with authentication"
}.
  Stack Trace:
     at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Primitives.ReferenceTypeAssertions`2.BeNull(String because, Object[] becauseArgs)
   at Surreal.Net.Tests.RpcClientTests.Signin() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 27
   at Surreal.Net.Tests.RpcClientTests.Use() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 34
   at Surreal.Net.Tests.RpcClientTests.Create() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 51
   at Surreal.Net.Tests.RpcClientTests.Change() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 82
   at Surreal.Net.Tests.RpcClientTests.Select() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 101
   at Surreal.Net.Tests.RpcClientTests.Query() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 116
--- End of stack trace from previous location ---
  Failed Surreal.Net.Tests.RpcClientTests.Select [6 ms]
  Error Message:
   Expected rsp.Error to be <null>, but found Surreal.Net.RpcError
{
    Code = -32000,
    Message = "There was a problem with authentication"
}.
  Stack Trace:
     at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Primitives.ReferenceTypeAssertions`2.BeNull(String because, Object[] becauseArgs)
   at Surreal.Net.Tests.RpcClientTests.Signin() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 27
   at Surreal.Net.Tests.RpcClientTests.Use() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 34
   at Surreal.Net.Tests.RpcClientTests.Create() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 51
   at Surreal.Net.Tests.RpcClientTests.Change() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 82
   at Surreal.Net.Tests.RpcClientTests.Select() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 101
--- End of stack trace from previous location ---
  Failed Surreal.Net.Tests.RpcClientTests.Use [8 ms]
  Error Message:
   Expected rsp.Error to be <null>, but found Surreal.Net.RpcError
{
    Code = -32000,
    Message = "There was a problem with authentication"
}.
  Stack Trace:
     at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message)
   at FluentAssertions.Execution.TestFrameworkProvider.Throw(String message)
   at FluentAssertions.Execution.DefaultAssertionStrategy.HandleFailure(String message)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(Func`1 failReasonFunc)
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args)
   at FluentAssertions.Primitives.ReferenceTypeAssertions`2.BeNull(String because, Object[] becauseArgs)
   at Surreal.Net.Tests.RpcClientTests.Signin() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 27
   at Surreal.Net.Tests.RpcClientTests.Use() in C:\Users\anton\Documents\Github\Surreal.Net\tests\Core.Tests\RpcTests.cs:line 34
--- End of stack trace from previous location ---

Failed!  - Failed:     8, Passed:     4, Skipped:     0, Total:    12, Duration: 283 ms - Core.Tests.dll (net6.0)
```

**Comment by ProphetLamb** [comment]
Hey, thanks a lot for your effort! The testcases were known to be erroneous, tho this should be fixed now 
 The Rest database doesn't support Authentication yet. In this sense, we expect the unit tests to fail.
The tests are the goal, that's TDD for you ^^

**Comment by ProphetLamb** [comment]
Feel free to open a pr anyway, I'll go about publishing a nuget package later today (after fixing up some issues) and your groundwork would be very much appreciated<3

**Comment by ProphetLamb** [comment]
On that note I'll also look into the errors you've found

Cheers

**Comment by ProphetLamb** [comment]
#11 Fixes most of the issues mentioned, one test - the REST client - still fails. That is due to a bad implementation.

**Comment by AntoniosBarotsis** [comment]
I pulled the updates and I'm also just missing one test now. Do you want me to also make GitHub releases publish to NuGet?

**Comment by ProphetLamb** [comment]
GitHub's workflows for tests would be awesome. I don't think that automated NuGet releases make sense before the initial release. (hopefully today). Feel free to implement whatever you like :)

**Comment by AntoniosBarotsis** [comment]
Great! 

I added the release workflow as well but feel free to delete it if you feel so (just undo the last commit). If you choose to keep it, make sure you create a repository secret `NUGET_API_KEY` with your NuGet API key. 

I would also recommend trying to run the publication once by hand to see that it works because you never know with these workflows.

```sh
dotnet build -c Release
dotnet pack ./src/Core/Core.csproj -c Release
dotnet nuget push ./src/Core/bin/Release/*.nupkg --api-key YOUR_API_KEY --source https://api.nuget.org/v3/index.json
```

Note that the version is currently picked up from [Directory.Build.props](https://github.com/ProphetLamb/Surreal.Net/blob/master/src/Directory.Build.props).

The publish workflow is largely copied from a project I made some time ago so in theory it should be fine.
